### PR TITLE
Prettifier & SourceInfo Refactoring

### PR DESCRIPTION
**Pull request by cheeseng**
Refactoring to enable support of customer Prettifier and use SourceInfo for stack depth resolving.

### CODE REVIEW

File: project/GenFactories.scala [hunk 1]
```diff
@@ -533,7 +532,7 @@ $endif$
    *                 ^
    * </pre>
    */
-  def and(containWord: ContainWord): AndContainWord = new AndContainWord
+  def and(containWord: ContainWord)(implicit prettifier: Prettifier, pos: source.Position): AndContainWord = new AndContainWord(prettifier, pos)
```
Comment (bvenners) [inline project/GenFactories.scala]: Why do and and or need the implicit prettifier and position here? Why can't the one captured by should be passed down?

File: common-test/src/main/scala/org/scalatest/SharedHelpers.scala [hunk 2]
```diff
@@ -828,7 +829,7 @@ object SharedHelpers extends Assertions with LineNumberHelper {
         case map: GenMap[_, _] => element.asInstanceOf[Tuple2[_, _]]._1
         case genTrv: GenTraversable[_] => getIndex(xs, element)
       }
-    Array(indexOrKey.toString, decorateToStringValue(element))
+    Array(indexOrKey.toString, decorateToStringValue(Prettifier.default, element))
```
Comment (bvenners) [inline common-test/src/main/scala/org/scalatest/SharedHelpers.scala]: Why is Prettifier.default used here? Why isn't a captured Prettifier passed down?

File: scalactic-macro/src/main/scala/org/scalactic/BooleanMacro.scala [hunk 3]
```diff
@@ -802,13 +802,13 @@ private[org] class BooleanMacro[C <: Context](val context: C, helperName: String
    *   [code generated from callHelper]
    * }
    */
-  def genMacro[T](booleanExpr: Expr[Boolean], methodName: String, clueExpr: Expr[Any]): Expr[T] = {
+  def genMacro[T](booleanExpr: Expr[Boolean], methodName: String, clueExpr: Expr[Any], prettifierExpr: Expr[_], posExpr: Expr[_]): Expr[T] = {
```
Comment (bvenners) [inline scalactic-macro/src/main/scala/org/scalactic/BooleanMacro.scala]: CS: Why Expr[_] for prettifierExpr and posExpr?

Comment (cheeseng) [inline scalactic-macro/src/main/scala/org/scalactic/BooleanMacro.scala]: The reason is that while BooleanMacro sits in scalactic-macro, Prettifier and Position sits in scalactic, which depends on scalactic-macro.  scalactic-macro has no access to Prettifier and Position at that point.

File: project/GenFactories.scala (hunk 1 above)
Comment (cheeseng) [inline project/GenFactories.scala]: I'll check back this one.

**Comment by cheeseng** [comment]
Majority changes are merged into 3.0.x already, the only one commit left is: 

https://github.com/scalatest/scalatest/pull/868/commits/f0c99b3a6448e26b4dbb4d406f4a4ca3665be6fe

We'll include this changes in a later PR.

**Pull request closed without merging**